### PR TITLE
Refactor Rewards parameters classes

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/37080

* Moves parameters access and storage logic into a single class (`RewardsParametersProvider`).
* Updates parameters pref storage so that there is a single `base::Value` pref that the parameters provider knows how to load and store. The parameters data in prefs no longer contains any "JSON strings inside of JSON".
* The helper registry has been updated to use `base::SupportsUserData`.
* The `PostChallenges` endpoint has been updated to follow new conventions.

Parameters pref before this change:

```json
{
  "ac": {
    "choice": 1.0,
    "choices": "[1.0,2.0,3.0,5.0,7.0,10.0,20.0]"
  },
  "payout_status": "{\"bitflyer\":\"processing\",\"gemini\":\"processing\",\"payoutDate\":\"\",\"solana\":\"processing\",\"unverified\":\"processing\",\"uphold\":\"processing\",\"zebpay\":\"\"}",
  "rate": 0.334756,
  "tip": {
    "choices": "[1.25,5.0,10.5]",
    "monthly_choices": "[1.25,5.0,10.5]"
  },
  "tos_version": 1,
  "vbat_deadline": "13343184000000000",
  "vbat_expired": true,
  "wallet_provider_regions": {
    "bitflyer": {
      "allow": ["..."],
      "block": []
    },
    "gemini": {
      "allow": ["..."],
      "block": []
    },
    "solana": {
      "allow": [],
      "block": ["..."]
    },
    "uphold": {
      "allow": ["..."],
      "block": []
    },
    "zebpay": {
      "allow": ["..."],
      "block": []
    }
  }
}
```

Parameters pref after this change:

```json
{
  "ac": {
    "choice": 1.0,
    "choices": [
      1.0,
      2.0,
      3.0,
      5.0,
      7.0,
      10.0,
      20.0
    ]
  },
  "payout_status": {
    "bitflyer": "processing",
    "gemini": "processing",
    "payoutDate": "",
    "solana": "processing",
    "unverified": "processing",
    "uphold": "processing",
    "zebpay": ""
  },
  "rate": 0.344127,
  "tip": {
    "choices": [
      1.25,
      5.0,
      10.5
    ],
    "monthly_choices": [
      1.25,
      5.0,
      10.5
    ]
  },
  "tos_version": 1,
  "vbat_deadline": "13343184000000000",
  "vbat_expired": true,
  "wallet_provider_regions": {
    "bitflyer": {
      "allow": [
        "..."
      ],
      "block": []
    },
    "gemini": {
      "allow": [
        "..."
      ],
      "block": []
    },
    "solana": {
      "allow": [],
      "block": [
        "..."
      ]
    },
    "uphold": {
      "allow": [
        "..."
      ],
      "block": []
    },
    "zebpay": {
      "allow": [
        "..."
      ],
      "block": []
    }
  }
}
```

No prefs migration has been added, since:

* The old format will be read in a best-effort manner, pulling in the fields that make sense.
* Parameters are refreshed on startup and periodically thereafter.

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by emerick — APPROVED** [review]

**Comment by bsclifton** [comment]
Added myself as a reviewer - just want to make sure I understand the changes here 😄

**Review by bsclifton — APPROVED** [review]
Got a chance to go through this today in-depth; LGTM!

Thanks for helping answer the questions I had over Slack😄  I like the updated pref storage (example in description made it crystal clear) and the `SupportsUserData` is a nice clean up too, IMO.

Only question I had; just came to mind. If someone had profile prefs in the old format and then they upgrade to this version, this would be handled gracefully, right? Basically the old pref is removed. And since those old values were just cached values, getting a new instance of the RewardsParameters and saving will update it.

**Comment by zenparsing** [comment]
@bsclifton 

> Only question I had; just came to mind. If someone had profile prefs in the old format and then they upgrade to this version, this would be handled gracefully, right? Basically the old pref is removed. And since those old values were just cached values, getting a new instance of the RewardsParameters and saving will update it.

Exactly. The parameters are queried on startup, so any JSON-in-JSON will be short-lived. If the endpoint can't be reached for a while, then the fields that contain JSON-in-JSON will just surface as default values. Fortunately, those values aren't critical.

File: components/brave_rewards/core/rewards_engine_helper.h [hunk 1]
```diff
@@ -40,6 +43,16 @@ class RewardsEngineHelper {
   const raw_ref<RewardsEngineImpl> engine_;
 };
 
+// A mixin for exposing a user data key for a `RewardsEngineHelper` class.
+template <typename T>
+class WithHelperKey {
```
Comment (bridiver) [inline components/brave_rewards/core/rewards_engine_helper.h:48]: What problem are you trying to solve here? I think this is making the rewards code harder for others to understand. If there's a generalized problem you think we need a solution for we should discuss and possibly submit to upstream.

**Pull request merged**